src/core/context-management/ContextManager.ts

```diff
@@ -9,57 +9,76 @@ import * as path from "path"
 
 // Type for the serialized format of our nested maps
 type SerializedContextHistory = Array<
 	[
 		number, // messageIndex
-		Array<
-			[
-				number, // blockIndex
-				ContextUpdate[], // updates array
-			]
-		>,
+		[
+			number, // EditType (message type)
+			Array<
+				[
+					number, // blockIndex
+					ContextUpdate[], // updates array (now with 4 elements including metadata)
+				]
+			>,
+		],
 	]
 >
 
 export class ContextManager {
 	// mapping from the apiMessages outer index to the inner message index to a list of actual changes, ordered by timestamp
 	// timestamp is required in order to support full checkpointing, where the changes we apply need to be able to be undone when
 	// moving to an earlier conversation history checkpoint - this ordering intuitively allows for binary search on truncation
+	// there is also a number stored for each (EditType) which defines which message type it is, for custom handling
 
-	// format:  {outerIndex => {innerIndex => [[timestamp, updateType, update], ...]}}
-	// example: { 1 => { 0 => [[<timestamp>, "text", "[NOTE] Some previous conversation history with the user has been removed ..."], ...] } }
+	// format:  { outerIndex => [EditType, { innerIndex => [[timestamp, updateType, update], ...] }] }
+	// example: { 1 => { [0, 0 => [[<timestamp>, "text", "[NOTE] Some previous conversation history with the user has been removed ..."], ...] }] }
 	// the above example would be how we update the first assistant message to indicate we truncated text
-	private contextHistoryUpdates: Map<number, Map<number, ContextUpdate[]>>
+	private contextHistoryUpdates: Map<number, [number, Map<number, ContextUpdate[]>]>
 
 	constructor() {
 		this.contextHistoryUpdates = new Map()
 	}
 
 	/**
-	 * public function for loading contextHistory from memory, if it exists
+	 * public function for loading contextHistoryUpdates from disk, if it exists
 	 */
 	async initializeContextHistory(taskDirectory: string) {
 		this.contextHistoryUpdates = await this.getSavedContextHistory(taskDirectory)
 	}
 
 	/**
-	 * get the stored context history from disk
+	 * get the stored context history updates from disk
 	 */
-	private async getSavedContextHistory(taskDirectory: string): Promise<Map<number, Map<number, ContextUpdate[]>>> {
+	private async getSavedContextHistory(taskDirectory: string): Promise<Map<number, [number, Map<number, ContextUpdate[]>]>> {
 		try {
 			const filePath = path.join(taskDirectory, GlobalFileNames.contextHistory)
 			if (await fileExistsAtPath(filePath)) {
 				const data = await fs.readFile(filePath, "utf8")
 				const serializedUpdates = JSON.parse(data) as SerializedContextHistory
 
-				return new Map(serializedUpdates.map(([messageIndex, innerMapArray]) => [messageIndex, new Map(innerMapArray)]))
+				// Update to properly reconstruct the tuple structure
+				return new Map(
+					serializedUpdates.map(([messageIndex, [numberValue, innerMapArray]]) => [
```

Comment:
essentially the same as above, number values here isn't descriptive enough for me to understand it's purpose.